pkg/knuu/instance_otel.go

```diff
@@ -0,0 +1,336 @@
+	}
+
+	if err := otelAgent.SetImage("otel/opentelemetry-collector-contrib:0.71.0"); err != nil {
+		return nil, fmt.Errorf("error setting image for otel-agent instance: %w", err)
+	}
+	if err := otelAgent.AddPortTCP(8888); err != nil {
+		return nil, fmt.Errorf("error adding port for otel-agent instance: %w", err)
+	}
+	if err := otelAgent.AddPortTCP(9090); err != nil {
+		return nil, fmt.Errorf("error adding port for otel-agent instance: %w", err)
+	}
+	if err := otelAgent.SetCPU("100m"); err != nil {
+		return nil, fmt.Errorf("error setting CPU for otel-agent instance: %w", err)
+	}
+	if err := otelAgent.SetMemory("100Mi", "200Mi"); err != nil {
+		return nil, fmt.Errorf("error setting memory for otel-agent instance: %w", err)
+	}
+	if err := otelAgent.Commit(); err != nil {
+		return nil, fmt.Errorf("error committing otel-agent instance: %w", err)
+	}
+
+	config := OTelConfig{
+		Extensions: i.createExtensions(),
+		Receivers:  i.createReceivers(),
+		Exporters:  i.createExporters(),
+		Service:    i.createService(),
+		Processors: i.createProcessors(),
+	}
+
+	bytes, err := yaml.Marshal(config)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling YAML: %w", err)
+	}
+
+	if err := otelAgent.AddFileBytes(bytes, "/etc/otel-agent.yaml", "root:root"); err != nil {
+		return nil, fmt.Errorf("error adding otel-agent config file: %w", err)
+	}
+
+	if err := otelAgent.SetCommand("/otelcol-contrib", "--config=/etc/otel-agent.yaml"); err != nil {
+		return nil, fmt.Errorf("error setting command for otel-agent instance: %w", err)
+	}
+
+	return otelAgent, nil
+}
+
+func (i *Instance) createExtensions() Extensions {
+	if i.otlpEndpoint == "" {
+		return Extensions{}
+	}
+
+	return Extensions{
+		BasicAuthOTLP: BasicAuthOTLP{
+			ClientAuth: ClientAuth{
+				Username: i.otlpUsername,
+				Password: i.otlpPassword,
+			},
+		},
+	}
+}
+
+func (i *Instance) createReceivers() Receivers {
+	receivers := Receivers{}
+
+	if i.otlpPort != 0 {
+		receivers.OTLP = OTLP{
+			Protocols: OTLPProtocols{
+				HTTP: OTLPHTTP{
+					Endpoint: fmt.Sprintf("localhost:%d", i.otlpPort),
+				},
+			},
+		}
+	}
+
+	if i.prometheusPort != 0 {
+		receivers.Prometheus = Prometheus{
+			Config: PrometheusConfig{
+				ScrapeConfigs: []ScrapeConfig{
+					{
+						JobName:        i.prometheusJobName,
+						ScrapeInterval: i.prometheusScrapeInterval,
+						StaticConfigs: []StaticConfig{
+							{
+								Targets: []string{fmt.Sprintf("localhost:%d", i.prometheusPort)},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	if i.jaegerGrpcPort != 0 {
+		receivers.Jaeger = Jaeger{
+			Protocols: JaegerProtocols{
+				GRPC: JaegerGRPC{Endpoint: fmt.Sprintf("localhost:%d", i.jaegerGrpcPort)},
+				ThriftCompact: JaegerThriftCompact{
+					Endpoint: fmt.Sprintf("localhost:%d", i.jaegerThriftCompactPort),
+				},
+				ThriftHTTP: JaegerThriftHTTP{
+					Endpoint: fmt.Sprintf("localhost:%d", i.jaegerThriftHttpPort),
+				},
+			},
+		}
+	}
+
+	return receivers
```

Comment:
for readibility purposes, I highly recommend to split this into funcs for each receiver we are setting for the overall receiver struct.

Reply:
https://github.com/celestiaorg/knuu/pull/174/commits/63b25751d3c9e9390afb8b154344d15d5396becb